src/Module.Server/Modes/Dtv/CrpgDtvClient.cs

```diff
@@ -36,13 +66,28 @@ protected override void AddRemoveMessageHandlers(GameNetwork.NetworkMessageHandl
         registerer.Register<CrpgDtvWaveStartMessage>(HandleWaveStart);
         registerer.Register<CrpgDtvViscountUnderAttackMessage>(HandleViscountUnderAttack);
         registerer.Register<CrpgDtvGameEnd>(HandleViscountDeath);
+        registerer.Register<CrpgDtvCurrentProgressMessage>(HandleCurrentProgress);
     }
 
     private void HandleSetTimer(CrpgDtvSetTimerMessage message)
     {
         TimerComponent.StartTimerAsClient(message.StartTime, message.Duration);
     }
 
+    private void HandleCurrentProgress(CrpgDtvCurrentProgressMessage message)
+    {
+        CurrentRound = message.Round + 1;
+        CurrentWave = message.Wave + 1;
+
+        Action onUpdateCurrentProgressEvent = OnUpdateCurrentProgress;
+        if (onUpdateCurrentProgressEvent == null)
+        {
+            return;
+        }
+
+        OnUpdateCurrentProgress();
+    }
+
```

Comment:
I'm not too familiar with race conditions , but correct me if i'm wrong but you would need to call
your local copy of the event "onUpdateCurrentProgressEvent" to prevent a race condition, right?
Here you assign it to check if it's null , yet call the original that may have become null meanwhile.

Anyway i've asked takeo , and except for the scripts where you can specify to have them parralelized,
all the game logic is in a singlethread.
So you should be able to safely do  OnUpdateCurrentProgress?.Invoke()

Reply:
I'll have a look at the proposed solution. I essentially copied this verbatim from some decompiled code.